### ISSUE TITLE
Receiving "Problem with applying template > The given key was not present in the dictionary." when applying template.

### ISSUE THREAD

**Issue by Sarah4x** [pnp/pnpframework#132]
This bug appears when I'm using App Only authentication and I'm trying to apply a provisioning template that includes ClientSidePages.

I've narrowed the bug down to an API call to..
GET /sites/<sitename>/_api/Web?$select=Id,Url,RegionalSettings&$expand=RegionalSettings/TimeZone

`
Fidder returns > 
HTTP/1.1 401 Unauthorized
Cache-Control: private
Server: Microsoft-IIS/10.0
x-ms-diagnostics: 3000006;reason="Token contains invalid signature.";category="invalid_client"
Strict-Transport-Security: max-age=31536000
X-FRAME-OPTIONS: SAMEORIGIN
SPRequestDuration: 24
SPIisLatency: 1
X-Powered-By: ASP.NET
MicrosoftSharePointTeamServices: 16.0.0.20913
X-Content-Type-Options: nosniff
X-MS-InvokeApp: 1; RequireReadOnly
P3P: CP="ALL IND DSP COR ADM CONo CUR CUSo IVAo IVDo PSA PSD TAI TELo OUR SAMo CNT COM INT NAV ONL PHY PRE PUR UNI"
WWW-Authenticate: Bearer realm=<token>
Date: Tue, 26 Jan 2021 16:07:32 GMT
Content-Length: 52

{"error_description":"Invalid issuer or signature."}`

And here's the stack trace..
"   
at System.Text.Json.JsonElement.GetProperty(String propertyName)\r\n   
at PnP.Core.SharePointRestError.ParseError(JsonElement error) in /_/src/sdk/PnP.Core/Exceptions/SharePointRestError.cs:line 65
   at PnP.Core.SharePointRestError..ctor(ErrorType type, Int32 httpResponseCode, String response) in /_/src/sdk/PnP.Core/Exceptions/SharePointRestError.cs:line 25
   at PnP.Core.SharePointRestServiceException..ctor(ErrorType type, Int32 httpResponseCode, String response) in /_/src/sdk/PnP.Core/Exceptions/SharePointRestServiceException.cs:line 33
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestInteractiveAsync>d__53.MoveNext() in /_/src/sdk/PnP.Core/Services/Core/BatchClient.cs:line 1255
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__39.MoveNext() in /_/src/sdk/PnP.Core/Services/Core/BatchClient.cs:line 262
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Model.BaseDataModel`1.<RequestAsync>d__101.MoveNext() in /_/src/sdk/PnP.Core/Model/Base/BaseDataModel.cs:line 908
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Services.PnPContextFactory.<InitializeContextAsync>d__39.MoveNext() in /_/src/sdk/PnP.Core/Services/Core/PnPContextFactory.cs:line 263
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__34.MoveNext() in /_/src/sdk/PnP.Core/Services/Core/PnPContextFactory.cs:line 186
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at PnP.Core.Services.PnPContextFactory.Create(Uri url, IAuthenticationProvider authenticationProvider) in /_/src/sdk/PnP.Core/Services/Core/PnPContextFactory.cs:line 165
   at PnP.Framework.PnPCoreSdk.GetPnPContext(ClientContext context) in /_/src/lib/PnP.Framework/PnPCoreSdk.cs:line 37

   at PnP.Framework.Provisioning.ObjectHandlers.ObjectClientSidePages.ProvisionObjects(Web web, ProvisioningTemplate template, TokenParser parser, ProvisioningTemplateApplyingInformation applyingInformation) in /_/src/lib/PnP.Framework/Provisioning/ObjectHandlers/ObjectClientSidePages.cs:line 41
   at PnP.Framework.Provisioning.ObjectHandlers.SiteToTemplateConversion.ApplyRemoteTemplate(Web web, ProvisioningTemplate template, ProvisioningTemplateApplyingInformation provisioningInfo, Boolean calledFromHierarchy, TokenParser tokenParser) in /_/src/lib/PnP.Framework/Provisioning/ObjectHandlers/SiteToTemplateConversion.cs:line 457"

**Comment by jansenbe** [comment]
@Sarah4x : I think you've an issue in your Azure AD app setup...can you cross check https://github.com/pnp/pnpframework/blob/dev/docs/MigratingFromPnPSitesCore.md#setting-up-an-azure-ad-app and verify your application was correctly setup.

Also, which version are you using? We did do some fixes in this space a while ago, so please ensure you're running the latest version

**Comment by Sarah4x** [comment]
@jansenbe I'm currently using the nightly, it happens on the latest stable too. I've cross checked with the migration document..

Azure AD is setup correctly for my own app. Made sure it had all the app-only permissions specified (I didn't have User.ReadWrite.All, but added it to be diligent). I should mention that my app is multi-tenant.

Old code to connect App Only >
`
new AuthenticationManager().GetAzureADAppOnlyAuthenticatedContext(webUrl, GlobalConfig.AzureADAppClientId, tenantId, cert);
`

New code >
`
var authManager = new AuthenticationManager(GlobalConfig.AzureADAppClientId, cert, tenantId);
return authManager.GetContext(webUrl);
`

I should mention that this code does work to connect to SharePoint, it just fails when being used within the ClientSidePages handler.

I saw this [issue](https://github.com/pnp/pnpframework/issues/86) which the last comment says he got it working, except when running locally, which I am doing too. 

Oh. And to be clear, I'm grabbing the cert by reading it from my local machine.
`var certPath = HostingEnvironment.MapPath("~/Certs/sarahs-custom-app.pfx");
                    var certfile = System.IO.File.OpenRead(certPath);
                    var certificateBytes = new byte[certfile.Length];
                    certfile.Read(certificateBytes, 0, (int)certfile.Length);
                    cert = new X509Certificate2(
                        certificateBytes,
                        "<password>",
                        X509KeyStorageFlags.Exportable |
                        X509KeyStorageFlags.MachineKeySet |
                        X509KeyStorageFlags.PersistKeySet);
                } `

and from the store in Azure in my App Service.
`var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
store.Open(OpenFlags.ReadOnly);
var certificates = store.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, false);
store.Close();
return certificates[0];`

This method has been working great in production for a year now using PnP-Sites-Core.

**Comment by jansenbe** [comment]
@Sarah4x : I've just re-ran my test code from the previous issue and below works for me. Can you: 

- Compare this snippet with your code, seems quite similar though
- When capturing the Fiddler trace, grab the access token from the Authorization header of the GET /sites//_api/Web?$select=Id,Url,RegionalSettings&$expand=RegionalSettings/TimeZone call and paste it in https://jwt.ms. You should see SharePoint roles in there:

![image](https://user-images.githubusercontent.com/7451219/105895275-9d071980-6015-11eb-9a35-c4089c27a9a0.png)

```csharp
            string certPath = @"C:\Users\bjansen\OneDrive - Microsoft\PnP\Certificates\BertOnlineAzureADAppOnlyDefaultPassword.pfx";
            string pwd = "xxx";

            var certfile = System.IO.File.OpenRead(certPath);
            var certificateBytes = new byte[certfile.Length];
            certfile.Read(certificateBytes, 0, (int)certfile.Length);
            var cert = new X509Certificate2(
                certificateBytes,
                pwd,
                X509KeyStorageFlags.Exportable |
                X509KeyStorageFlags.MachineKeySet |
                X509KeyStorageFlags.PersistKeySet);

            using (var clientContext = new AuthenticationManager("appid", cert, "bertonline.onmicrosoft.com")
                                                         .GetContext("https://bertonline.sharepoint.com/sites/TheLanding1"))
            {

                // Basic test to see if we've SharePoint access
                //clientContext.Load(clientContext.Web, p => p.Title);
                //clientContext.ExecuteQueryRetry();

                // This triggers the call out to PnP Core SDK
                //var pnpContext = PnPCoreSdk.Instance.GetPnPContext(clientContext);
                //var pages = pnpContext.Web.GetPages("Home.aspx");

                ProvisioningTemplateApplyingInformation ptai = new ProvisioningTemplateApplyingInformation()
                {
                    //HandlersToProcess = Handlers.PageContents,
                    ProgressDelegate = delegate (String message, Int32 progress, Int32 total)
                    {
                        // Only to output progress for console UI
                        Console.WriteLine("{0:00}/{1:00} - {2}", progress, total, message);
                    }
                };

                XMLTemplateProvider provider = new XMLFileSystemTemplateProvider(@"d:\temp\modernpages", "");
                ProvisioningTemplate sourceTemplate = provider.GetTemplate("yannick.xml");
                sourceTemplate.Connector = new FileSystemConnector(@"d:\temp\modernpages", "");

                // Execute actual extraction of the tepmplate
                clientContext.Web.ApplyProvisioningTemplate(sourceTemplate, ptai);                
            }
```

The used app is multi-tenant and has these permissions:
![image](https://user-images.githubusercontent.com/7451219/105893832-dd659800-6013-11eb-8f67-a1219e7d7f8c.png)

The used template is this one:

```Xml
<?xml version="1.0"?>
<pnp:Provisioning xmlns:pnp="http://schemas.dev.office.com/PnP/2020/02/ProvisioningSchema">
  <pnp:Preferences Generator="PnP.Framework, Version=0.1.101.0, Culture=neutral, PublicKeyToken=null" />
  <pnp:Templates ID="CONTAINER-TEMPLATE-597237A407A441458763A310FA201190">
    <pnp:ProvisioningTemplate ID="TEMPLATE-597237A407A441458763A310FA201190" Version="1" BaseSiteTemplate="SITEPAGEPUBLISHING#0" Scope="RootSite">
      <pnp:ClientSidePages>
        <pnp:ClientSidePage PromoteAsNewsArticle="false" PromoteAsTemplate="true" Overwrite="true" Title="Article - Advanced Layout" ThumbnailUrl="" PageName="StickyTemplate.aspx">
          <pnp:Header Type="Custom" ServerRelativeImageUrl="{site}/SiteAssets/SitePages/newpage/40298-NewPerspective-newspost.jpg" LayoutType="ColorBlock" ShowTopicHeader="true" ShowPublishDate="false" TopicHeader="NEWS" AlternativeText="" Authors="[]" AuthorByLine="[]" AuthorByLineId="-1" />
          <pnp:Sections>
            <pnp:Section Order="1" Type="TwoColumnLeft">
              <pnp:Controls>
                <pnp:CanvasControl WebPartType="Text" ControlId="70148360-90cb-4ae9-8e56-4d5f7e02b186" Order="1" Column="1">
                  <pnp:CanvasControlProperties>
                    <pnp:CanvasControlProperty Key="Text" Value="A sticky description..." />
                  </pnp:CanvasControlProperties>
                </pnp:CanvasControl>
              </pnp:Controls>
            </pnp:Section>
          </pnp:Sections>
        </pnp:ClientSidePage>
      </pnp:ClientSidePages>
    </pnp:ProvisioningTemplate>
  </pnp:Templates>
</pnp:Provisioning>
```

**Comment by Sarah4x** [comment]
@jansenbe I figured it out! Thanks for all your help.

I was using a Provisioning Context and I had to update the AcquireTokenAsyncDelegate.

I ended up instantiating my Authentication Manager in a separate method and then using the GetAccessTokenAsync with the web url and it worked!

`
using (var provisioningContext = new PnPProvisioningContext((resource, scope) => { return Connect.GetAppOnlyAuthManager(tenantGuid).GetAccessTokenAsync(webUrl);}))
                        {
                            //#endif
                            cc.Web.ApplyProvisioningTemplate(provisioningTemplate, applyingInformation);
}`

Thanks again for your help!
